write-fonts/src/tables/cmap.rs

```diff
@@ -71,21 +71,29 @@ impl Cmap {
 
         let mut prev = (u16::MAX - 1, u16::MAX - 1);
         for (cp, gid) in mappings.into_iter() {
+            let gid = gid.to_u16();
             let cp = (cp as u32).try_into().unwrap();
             let next_in_run = (
                 prev.0.checked_add(1).unwrap(),
                 prev.1.checked_add(1).unwrap(),
             );
-            let current = (cp, gid.to_u16());
+            let current = (cp, gid);
             // Codepoint and gid need to be continuous
             if current != next_in_run {
                 // Start a new run
                 start_code.push(cp);
                 end_code.push(cp);
 
-                // we might need to reach further than an i16 can take us
-                // using idRangeOffset ... but we're saving that for another day
-                id_deltas.push((gid.to_u16() as i32 - cp as i32).try_into().unwrap());
+                // TIL Python % 0x10000 and Rust % 0x10000 do not mean the same thing.
+                // rem_euclid is almost what we want, except as applied to small values
+                // ex -10 rem_euclid 0x10000 = 65526
```

Comment:
Hm fonttools uses _unsigned_ uint16 for idDeltas, so just doing `(indices[i] - startCode[i]) % 0x10000` always works even for very small indices as in your example above -10 % 0x10000 == 65526:

https://github.com/fonttools/fonttools/blob/3b9a73ff8379ab49d3ce35aaaaf04b3a7d9d1655/Lib/fontTools/ttLib/tables/_c_m_a_p.py#L778
note the use of "H" for initializing array.array, which means unsigned short integer (lowercase "h" would mean signed):
https://github.com/fonttools/fonttools/blob/3b9a73ff8379ab49d3ce35aaaaf04b3a7d9d1655/Lib/fontTools/ttLib/tables/_c_m_a_p.py#L1018
there is how it obtains back a gid by summing up (charCode + idDelta) % 0xFFFF, in the subtable4 decompile method:  
https://github.com/fonttools/fonttools/blob/3b9a73ff8379ab49d3ce35aaaaf04b3a7d9d1655/Lib/fontTools/ttLib/tables/_c_m_a_p.py#L913

However, you're right that the spec uses int16:
https://learn.microsoft.com/en-us/typography/opentype/spec/cmap#format-4-segment-mapping-to-delta-values

I wonder if it used to be uint16 at some point in the past, then the spec was updated later on..
In any case, I don't think it matters the difference, since those idDeltas are not exposed to the public API in fonttools and only used in the compile/decompile steps. We seem to be getting the same numbers in the end, right?

Reply:
look what I found: https://github.com/fonttools/fonttools/issues/2575

Reply:
interestingly, in the linked discussion, https://github.com/MicrosoftDocs/typography-issues/issues/980, Peter Constable notes

> this would have been made easier if the type for idDelta were kept as unsigned as it was in the original TrueType spec...

Reply:
I see when read-fonts decompiles the glyph index it performs a cast to u16 immediately after summing codepoint and the (signed i16) delta, like this `(codepoint as i32 + delta) as u16`
https://github.com/googlefonts/fontations/blob/1f85b76a4697c03221388cfd430fff9b4664a3b5/read-fonts/src/tables/cmap.rs#L61

So in the end it doesn't matter whether we interpet idDeltas as signed (as OT spec and fontations do) or as unsigned (as fonttools and old TrueType spec do), the end result is the same.

Reply:
I didn't realize it had changed signedness, thats amazing.